Show a welcome screen if no database is open

oops-shlok opened the pull request:
Closes #12272 

### Mandatory checks

<!-- 
- Go through the list below. Please don't remove any items.
- [x] done; [ ] not done / not applicable
-->

- [x] I own the copyright of the code submitted and I licence it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if change is visible to the user)
- [ ] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [ ] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.



https://github.com/user-attachments/assets/20f73b25-0f3b-4863-b992-854b6312cf8f

<img width="1626" alt="Screenshot 2025-02-04 at 3 49 48 PM" src="https://github.com/user-attachments/assets/d4297b29-8740-4bb7-88dc-9eea93496af6" />
<img width="1626" alt="Screenshot 2025-02-04 at 3 48 51 PM" src="https://github.com/user-attachments/assets/a002d509-e5b8-4bcb-9b9d-544d350f7d07" />

calixtus commented on src/main/java/org/jabref/gui/Base.css:
```diff
@@ -1734,9 +1734,41 @@ We want to have a look that matches our icons in the tool-bar */
+    -fx-font-family: "Arial";
+}
+
+.description-label {
```
Please use a more explizit selector name for the CSS class, as this may interfere with other places similar or the same classes are used. Use welcome-* like above and ahead as a prefix.

calixtus commented on src/main/java/org/jabref/gui/Base.css:
```diff
@@ -1734,9 +1734,41 @@ We want to have a look that matches our icons in the tool-bar */
 /* AboutDialog */
 #aboutDialog .about-heading {
-    -fx-font-size: 30;
+    -fx-font-size: 31;
```
Why this change?

oops-shlok replied:
Reverted this change

calixtus commented on src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -230,11 +258,27 @@ private void initLayout() {
         setTop(head);
 
         splitPane.getItems().addAll(tabbedPane);
+        contentPane.getChildren().addAll(welcomePage, splitPane);
```
I would suggest to implement this as a separate tab and not as a static background. Take vscode as an example.
This way, the welcome page can be kept shown and revisited later if the user wants to work on a library in parallel. I think implementation would be straightforward as a new tab.
This would also take a lot of complexity out of the node tree.

oops-shlok replied:
I have implemented a WelcomeTab now and tried to implement it how its done for vscode. I have also handled the save, save as, export etc. edge cases with respect to welcomeTab. Please let me know if this is fine or changes are needed for this.

calixtus commented on src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -230,11 +258,27 @@ private void initLayout() {
+
+    private void updateContent() {
+        boolean hasOpenDatabases = !stateManager.getOpenDatabases().isEmpty();
+        welcomePage.setVisible(!hasOpenDatabases);
```
If the welcome page is the most 'lowest' node in a StackPane, it should automatically be hidden if a library is shown above. This is the whole purpose of a StackPane to deal with this. So there should be no need for messing with visibility stuff here...

calixtus commented on src/main/java/org/jabref/gui/util/WelcomePage.java:
```diff
@@ -0,0 +1,124 @@
+import org.jabref.model.entry.BibEntryTypesManager;
+import org.jabref.model.util.FileUpdateMonitor;
+
+public class WelcomePage extends VBox {
```
Extends Tab?

oops-shlok replied:
Extends Tab is now there for WelcomeTab

calixtus commented on src/main/java/org/jabref/gui/util/WelcomePage.java:
```diff
@@ -0,0 +1,124 @@
+        welcomeLabel.getStyleClass().add("welcome-label");
+
+        Label descriptionLabel = new Label(Localization.lang("Stay on top of your Literature"));
+        descriptionLabel.getStyleClass().add("description-label");
```
Explicit class names

calixtus commented on src/main/java/org/jabref/gui/util/WelcomePage.java:
```diff
@@ -0,0 +1,124 @@
+        return createBox(Pos.TOP_LEFT, welcomeLabel, descriptionLabel);
+    }
+
+    private VBox createStartBox(JabRefFrame frame, GuiPreferences preferences, AiService aiService, DialogService dialogService, StateManager stateManager, FileUpdateMonitor fileUpdateMonitor, BibEntryTypesManager entryTypesManager, CountingUndoManager undoManager, ClipBoardManager clipBoardManager, TaskExecutor taskExecutor) {
```
Why not class variables? I don't see any need for method injection.

calixtus commented on src/main/java/org/jabref/gui/util/WelcomePage.java:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    private VBox createBox(Pos alignment, Node... nodes) {
+        VBox box = new VBox(5);
```
Why fixed count for initialization?

oops-shlok replied:
This is for spacing when VBox will be initialised.

calixtus commented on src/main/java/org/jabref/gui/util/WelcomePage.java:
```diff
@@ -0,0 +1,124 @@
+        }
+    }
+
+    private VBox createBox(Pos alignment, Node... nodes) {
```
CreateAlignedVBox?

oops-shlok replied:
specified alignment as TOP_LEFT for the all the VBox now.

calixtus reviewed (changes requested):
Just a quick review from my cellphone. A more deeper look will follow.

oops-shlok commented:
Hi @calixtus! Saw all the comments with respect to the PR. Will resolve them asap. Thanks for the review!

subhramit commented on buildres/abbrv.jabref.org:
Undo the changes to submodules. Follow the alternative method in https://devdocs.jabref.org/code-howtos/faq.html#submodules

oops-shlok replied:
Noted. Will do this from now. Reverted the changes to submodules.

subhramit commented on src/main/resources/csl-styles:
Same as above

oops-shlok replied:
Reverted the changes to submodules

subhramit reviewed (dismissed):
Hey @oops-shlok, thanks for your contribution!
Please revert the submodule changes (steps linked below).

Also, please don't resolve conversations from your end until the reviewer who asked for the change is satisfied and does that - otherwise each comment has to be manually expanded when trying to gather context on the changes asked (when re-reviewing) or to read your replies and carry the discussion forward.
For very trivial comments (for example, these ones below), you can mark resolved once done.

koppor commented:
For reference, the current VSCode welcome screen offers walkthroughs:  - Maybe as follow-up

![image](https://github.com/user-attachments/assets/2b8a1319-438a-4861-90e2-4e24e09a8944)

koppor commented on src/main/java/org/jabref/gui/LibraryTab.java:
```diff
@@ -809,6 +809,11 @@ private void onClosed(Event event) {
             tableModel.unbind();
         }
+
+        if (stateManager.getOpenDatabases().isEmpty() && stateManager.activeDatabaseProperty().isPresent().get()) {
```
This check is new.

I miss a "why". I think, it is because of the welcome screen. Then, please, add it as comment.

    // This check is needed, because the WelcomeTab is a tab, but not a database.

koppor commented on src/main/java/org/jabref/gui/WelcomeTab.java:
```diff
@@ -0,0 +1,146 @@
+                      TaskExecutor taskExecutor,
+                      FileHistoryMenu fileHistoryMenu) {
+
+        super("Welcome");
```
I think, there is a Localization missing?

koppor commented on src/main/java/org/jabref/gui/WelcomeTab.java:
```diff
@@ -0,0 +1,146 @@
+        Label welcomeLabel = new Label(Localization.lang("Welcome to JabRef"));
+        welcomeLabel.getStyleClass().add("welcome-label");
+
+        Label descriptionLabel = new Label(Localization.lang("Stay on top of your Literature"));
```
```suggestion
        Label descriptionLabel = new Label(Localization.lang("Stay on top of your literature"));
```

koppor commented on src/main/java/org/jabref/gui/importer/actions/OpenDatabaseAction.java:
```diff
@@ -147,10 +147,15 @@ public FileDialogConfiguration getFileDialogConfiguration(Path initialDirectory)
         }
+
+        LibraryTab currentTab = tabContainer.getCurrentLibraryTab();
+        if (currentTab == null) {
```
Please add a Java comment that this is because of the `WelcomeTab` class.

koppor commented on src/main/resources/l10n/JabRef_en.properties:
```diff
@@ -2854,3 +2854,11 @@ Include=Include
 Include\ or\ exclude\ cross-referenced\ entries=Include or exclude cross-referenced entries
 Would\ you\ like\ to\ include\ cross-reference\ entries\ in\ the\ current\ operation?=Would you like to include cross-reference entries in the current operation?
+
+New\ Library=New Library
```
Very strange that a new localization is needed.

Please rework the strings to be in line with the other casing in JabRef

![grafik](https://github.com/user-attachments/assets/2bac884a-d3aa-4d0f-a0e0-6c12b35fa0c6)

calixtus commented:
Hi, I finally tested your changes in JabRef. I like it very much and I think its a great addition for new users with a big impact. I have two small suggestions from testing it:

1. The welcomeTab should only open, if there is not already another file opened from last time (check for OpenDatabases object in `uiCommands`).
2. Add a menu entry in help "open welcome tab".

Thanks!

calixtus commented:
Do not force-push! Force push is off limits and very bad style when working together on a project. Commits will be lost, comments on commits will loose their context. This makes it harder to review. In the end, all commits will be squished either way before being merged into the main branch.

oops-shlok commented:
Sure @calixtus & @koppor ! I will keep this in mind. Actually due to some recent changes in the JabRefFrame class, my code is affected. So, I have to fix that according to the new code in JabRefFrame class. Also, in this process, there is some issue not with commits and branches. So, if you are fine with it then can I make a new PR from a new branch?

calixtus commented:
No. Stay with this pr. just merge upstream/main into your branch in your local machine and then push it to the remote (github) repo

calixtus commented:
This is part of learning how git works, to work with changing merge targets and to keep your pr up-to-date. You can text us in the gitter chat if you are stuck and don't understand something https://app.gitter.im/#/room/#JabRef_jabref:gitter.im

oops-shlok commented:
Understood @calixtus. I have fixed the issue now. This branch now aligns with the main branch along with my changes. Now I will implement your suggestions i.e: 

> Hi, I finally tested your changes in JabRef. I like it very much and I think its a great addition for new users with a big impact. I have two small suggestions from testing it:
> 
> 1. The welcomeTab should only open, if there is not already another file opened from last time (check for OpenDatabases object in `uiCommands`).
> 2. Add a menu entry in help "open welcome tab".
> 
> Thanks!

oops-shlok commented:
Hi @koppor & @calixtus! Implemented the functionality for the suggestions you gave. Please review once and lmk if this is fine.

koppor commented on src/main/java/org/jabref/gui/actions/StandardActions.java:
```diff
@@ -207,7 +208,6 @@ public enum StandardActions implements Action {
     GROUP_ENTRIES_REMOVE(Localization.lang("Remove selected entries from this group")),
 
     CLEAR_EMBEDDINGS_CACHE(Localization.lang("Clear embeddings cache"));
-
```
I think, the empty line should be kept.

koppor commented on src/main/java/org/jabref/gui/exporter/ExportCommand.java:
```diff
@@ -133,19 +133,23 @@ private void export(Path file, FileChooser.ExtensionFilter selectedExtensionFilt
-                                notificationPane.hide();
-                            })),
-                            Duration.seconds(5));
+                    boolean libraryTab = tabSupplier.get() != null;
```
The boolean variable should be named `currentTabIsLibraryTab`.

@calixtus Should we introduce an `Optional` at the `get()`? - Or do we need some more sophisticated logic here? Maybe also @InAnYan

koppor commented on src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -171,10 +183,6 @@ public JabRefFrame(Stage mainStage,
-                this.preferences.getLastFilesOpenedPreferences().getFileHistory(),
-                dialogService,
-                getOpenDatabaseAction());
         this.setOnKeyTyped(key -> {
             if (this.fileHistory.isShowing()) {
```
Why wasn't this moved up? All functionality of the fileHistory menu should be kept togeher.

koppor commented on src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -122,6 +124,16 @@ public JabRefFrame(Stage mainStage,
         this.clipBoardManager = clipBoardManager;
         this.taskExecutor = taskExecutor;
 
+        this.fileHistory = new FileHistoryMenu(
+                this.preferences.getLastFilesOpenedPreferences().getFileHistory(),
+                dialogService,
+                getOpenDatabaseAction());
```
Why the move before `setId`?

Shouldn't it be below `// Create components`?

koppor commented on src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -225,12 +233,42 @@ private void initLayout() {
         setTop(head);
 
-        splitPane.getItems().addAll(tabbedPane);
+        boolean hasPreviousFiles = !stateManager.getOpenDatabases().isEmpty();
+
+        if (!hasPreviousFiles) {
```
Merge into one line. The variable is not used further down, is it?

(The variable name is pretty bad - it should be `openedDatabasesExisting`.

koppor commented on src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -428,16 +466,39 @@ private void updateTabBarVisible() {
      * Returns the currently viewed LibraryTab.
      */
     public LibraryTab getCurrentLibraryTab() {
```
i think, this sshould be en Optional - don't know, how much effort this is.

calixtus replied:
This is way beyond the scope of this pr. I checked the changes made, this is a bottomless pit.

koppor commented on src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -428,16 +466,39 @@ private void updateTabBarVisible() {
     }
 
+    public void showWelcomeTab() {
+        for (Tab tab : tabbedPane.getTabs()) {
+            if (!(tab instanceof LibraryTab)) {
+                tabbedPane.getSelectionModel().select(tab);
+                return;
+            }
```
Why this loop? There should be a Java comment why this is needed.

Maybe also JavaDoc to showWelcomeTab that it is only shown of no other tabs exist. -- Update: I think, you are checking for WelcomeTab - why is the `if` not checking for instanceOf WelcomeTab?!

koppor commented on src/main/resources/l10n/JabRef_en.properties:
```diff
@@ -2854,5 +2854,14 @@ Include=Include
 Entries\ copied\ successfully,\ including\ cross-references.=Entries copied successfully, including cross-references.
 Entries\ copied\ successfully,\ without\ cross-references.=Entries copied successfully, without cross-references.
+
+No\ Recent\ Libraries=No Recent Libraries
```
```suggestion
No\ recent\ libraries=No recent libraries
```

JabRef uses lower case:

![grafik](https://github.com/user-attachments/assets/ce2ee822-3e77-4e3c-bdda-c67c10ff3d88)

oops-shlok commented:
> > Hi @koppor & @calixtus! Implemented the functionality for the suggestions you gave. Please review once and lmk if this is fine.
> 
> Since the PR is now two eeks in progress - did you change something in the screen or are the screenshots in the PR description still valid? -- In either case, please report. (And in the future, please try to comment whether the screenshots are still valid)

Hi @koppor! Now the welcome screen is implemented as a tab. I will add the new screenshots after I implement the changes suggested by you.

oops-shlok commented:
Hi @koppor & @calixtus ! Implemented the changes suggested by @koppor. Please review the changes and lmk if this is fine. Also, added the new screenshots after the changes. 

<img width="1626" alt="Screenshot 2025-02-19 at 2 06 48 AM" src="https://github.com/user-attachments/assets/2fe52ed0-3794-4985-886c-f50cb29b5020" />
<img width="1625" alt="Screenshot 2025-02-19 at 2 08 00 AM" src="https://github.com/user-attachments/assets/5c20f40b-8291-4c53-a181-f7dc0f2cc03a" />
<img width="1623" alt="Screenshot 2025-02-19 at 2 09 05 AM" src="https://github.com/user-attachments/assets/36dc6f1c-d15d-4fde-889c-41ee159e62ac" />

oops-shlok commented:
Hi @koppor @calixtus ! Any suggestions for the updated code?

calixtus commented on src/main/java/org/jabref/gui/LibraryTab.java:
```diff
@@ -809,6 +809,12 @@ private void onClosed(Event event) {
         }
+
+        // This check is needed, because the WelcomeTab is a tab, but not a database.
+        if (stateManager.getOpenDatabases().isEmpty() && stateManager.activeDatabaseProperty().isPresent().get()) {
+            stateManager.activeDatabaseProperty().setValue(Optional.empty());
+        }
```
I don't get why this check is in LibraryTab::onClose . This seems very very odd to me. Shouldn't this go to the "poor-mans" binding in JabRefFrame?

calixtus commented on src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -330,15 +366,15 @@ private void initBindings() {
         EasyBind.subscribe(stateManager.activeSearchQuery(SearchType.NORMAL_SEARCH), query -> {
-            if (getCurrentLibraryTab() != null) {
-                getCurrentLibraryTab().searchQueryProperty().set(query);
+            if (getCurrentLibraryTab().isPresent()) {
+                getCurrentLibraryTab().ifPresent(tab -> tab.searchQueryProperty().set(query));
```
isPresent - if Present? This does not make any sense at all. Please do not just accept all the changes the IDE suggests.

koppor commented on src/main/java/org/jabref/gui/LibraryTabContainer.java:
```diff
@@ -13,7 +14,7 @@ public interface LibraryTabContainer {
     List<LibraryTab> getLibraryTabs();
 
     @Nullable
-    LibraryTab getCurrentLibraryTab();
+    Optional<LibraryTab> getCurrentLibraryTab();
```
The suggestion was by me - the code inside should return an empty optional if there is no curent library tab.

oops-shlok commented:
Hmm..Sure. I will then revert the changes done to the interface.

calixtus commented:
Please excuse the inconvenience, we did not initially realize how much work would result from this small change in the interface. This would go way beyond of the scope of this PR about the WelcomeTab. We should postpone that change to a followup pr eventually. Will you do the change back or should I do the chores? 😄

calixtus commented:
Hi, how is your progress? I really like that feature and would love to have it included?

oops-shlok commented:
Hi @calixtus , @koppor! The changes are reverted and the save and save as functionality is fixed according to the review suggestion. Please review once and lmk if any changes are needed.

koppor commented:
Current state:

![image](https://github.com/user-attachments/assets/b56eb810-3525-4cc4-8c1f-ae9628957a15)

---

Optional tasks:

- some spacing on the left if space is too small
- smaller font size
- Work if library was closed and tab bar is hidden

I think, it does not work if the tab bar is hidden:

![image](https://github.com/user-attachments/assets/54af970c-6523-4792-973b-54886450dffd)

![image](https://github.com/user-attachments/assets/def648c2-cc0b-4d86-bdc6-ea2f9be20f80)

koppor reviewed (dismissed):
Looks good in normal settings.

Looks bad in shrinked JabRef, does not open single library is closed.

I would like to have the weclome tab be closed as soon as a library is opened, but this is a gut feeling only.

It is a good step forward, thus OK for me to merge this in and fix in follow-up PRs.

calixtus commented:
I will push some more commits, we can work together if you want, @oops-shlok, but I can also work on the remaining issues myself. Got some time tonight and tomorrow

oops-shlok commented:
> I will push some more commits, we can work together if you want, @oops-shlok, but I can also work on the remaining issues myself. Got some time tonight and tomorrow

Hi @calixtus! Sure. Lets work on the remaining issues. Shall I look into the issue that @koppor mentioned?

calixtus commented:
Yes, i will look at the issue about hidden tab bar, you can look into issue about small window size.
Remember to update your local branch by merging (not rebasing!) frequently, so we stay in sync with our branches

oops-shlok commented:
Sure @calixtus! On it.

oops-shlok commented:
hi @koppor & @calixtus! So, I have implemented the spacing and smaller font size for smaller windows. Can you please check and lmk if this is fine? Attaching the screenshots for your reference. 

For smaller window:

<img width="649" alt="Screenshot 2025-03-07 at 11 54 54 PM" src="https://github.com/user-attachments/assets/d9502ce2-4c2f-4e2b-a589-ea805e2ea61a" />

For normal window:

<img width="1626" alt="Screenshot 2025-03-07 at 11 54 11 PM" src="https://github.com/user-attachments/assets/1f7d8e8e-7d3d-48af-8e02-ae15dffdc177" />

For normal window with web search:

<img width="1626" alt="Screenshot 2025-03-07 at 11 54 03 PM" src="https://github.com/user-attachments/assets/1171c6cc-7c70-4d81-aafc-ed71d18ccaf3" />

calixtus commented:
Please push your commits

koppor commented:
> Yes, i will look at the issue about hidden tab bar

Maybe it is happening even with non-hidden tab bar: If the last library tab is closed, the welcome screen is not opened again.

calixtus commented:
> > Yes, i will look at the issue about hidden tab bar
> 
> Maybe it is happening even with non-hidden tab bar: If the last library tab is closed, the welcome screen is not opened again.

Yes, this is the desired state. WelcomeTab should only open, when no library is open on start, as the example vscode does. Reasoning: We assume that a first time user needs some guidance how to use JabRef and likes to be welcomed. We assume that power users find the welcome tab annoying. Thus if the user really wants to reopen the WelcomeTab, he or she can do so manually by calling it via the help menu.
A followup PR could integrate a checkbox "do not show again". But lets focus on finishing this iteration first.

koppor commented:
> Yes, this is the desired state. WelcomeTab should only open, when no library is open on start, as the example vscode does. 

I checked: VS Code behaves the same. Thus, I agree.

Side note: if no files are opened after closing the welcome tab, VS Code shows a hint on keyboard short cuts:

![image](https://github.com/user-attachments/assets/8a286c2e-c294-406b-993c-6d01b81141fb)

> Thus if the user really wants to reopen the WelcomeTab, he or she can do so manually by calling it via the help menu. 

Works. Thank you for thinking of that!

> A followup PR could integrate a checkbox "do not show again". But lets focus on finishing this iteration first.

Thanks to your explanation, it is OK for me!

koppor commented:
> For smaller window:

The font size is larger than for the menu font size. It is possible to reduce it?

For the larger Windows, it is OK.

Note that it OK if this is possible with much effort only.

oops-shlok commented:
> Is it easily possible to add following to the tab - or should I open a new issue?
> 
> ## Community
> * [Online help](https://docs.jabref.org/)
> * [Forum for support](https://discourse.jabref.org/)
> * [Mastodon](https://mastodon.acm.org/@jabref@foojay.social)
> * [LinkedIn](https://linkedin.com/company/jabref/)
> * [Development version](https://builds.jabref.org/main/)
> * [CHANGELOG](https://github.com/JabRef/jabref/blob/main/CHANGELOG.md)
> * [Donation](https://donations.jabref.org)
> 
> Use same icons as in "Help" -> "JabRef resources"

Sure, I will add this in the welcome page

oops-shlok commented:
> > Is it easily possible to add following to the tab - or should I open a new issue?
> > ## Community
> > 
> > * [Online help](https://docs.jabref.org/)
> > * [Forum for support](https://discourse.jabref.org/)
> > * [Mastodon](https://mastodon.acm.org/@jabref@foojay.social)
> > * [LinkedIn](https://linkedin.com/company/jabref/)
> > * [Development version](https://builds.jabref.org/main/)
> > * [CHANGELOG](https://github.com/JabRef/jabref/blob/main/CHANGELOG.md)
> > * [Donation](https://donations.jabref.org)
> > 
> > Use same icons as in "Help" -> "JabRef resources"
> 
> Sure, I will add this in the welcome page

Hi @koppor  & @calixtus! any suggestions for the UI of this section. For now, I have added this as a footer in the welcome tab. But, I feel the it can be made better. Could you please suggest or provide me any reference for this section? 
<img width="1624" alt="Screenshot 2025-03-11 at 12 38 53 AM" src="https://github.com/user-attachments/assets/b4599bb6-d767-42ce-908c-2033dac12b2c" />

koppor commented:
Change text to "download development version":

![image](https://github.com/user-attachments/assets/796065d1-4cc5-456a-9ed6-dfd62f8f9670)

It should additionally output the current version:

![image](https://github.com/user-attachments/assets/c6b35d87-5b2e-4c1f-8788-998571896576)

oops-shlok commented:
Hi @koppor & @calixtus! Please take a look at the updated UI and lmk if any changes are needed. Attaching the screenshot.

<img width="1624" alt="Screenshot 2025-03-12 at 1 34 39 AM" src="https://github.com/user-attachments/assets/e4fac5be-e442-44a2-a689-0f18f062a591" />

koppor commented:
> Hi @koppor & @calixtus! Please take a look at the updated UI and lmk if any changes are needed. Attaching the screenshot.

You did not push anything, did you?

We sould like to get this in.

Maybe we merge as is and then follow-ups can be made.

oops-shlok commented:
> > Hi @koppor & @calixtus! Please take a look at the updated UI and lmk if any changes are needed. Attaching the screenshot.
> 
> You did not push anything, did you?
> 
> We sould like to get this in.
> 
> Maybe we merge as is and then follow-ups can be made.

Sure! will push the changes right away.

oops-shlok commented:
Hi @koppor & @calixtus! Is there something that still needs to be fixed for the welcome tab? If there is anything, please lmk so that I can work on it asap.

koppor commented on src/main/java/org/jabref/gui/WelcomeTab.java:
```diff
@@ -0,0 +1,250 @@
+    private void updateWelcomeRecentLibraries() {
+        recentLibrariesBox.getChildren().clear();
+
+        if (fileHistoryMenu.getItems().isEmpty()) {
+            Label noRecentLibrariesLabel = new Label(Localization.lang("No recent libraries"));
+            noRecentLibrariesLabel.getStyleClass().add("welcome-no-recent-label");
+            recentLibrariesBox.getChildren().add(noRecentLibrariesLabel);
+        } else {
+            for (MenuItem item : fileHistoryMenu.getItems()) {
+                Hyperlink recentLibraryLink = new Hyperlink(item.getText());
+                recentLibraryLink.getStyleClass().add("welcome-hyperlink");
+                recentLibraryLink.setOnAction(item.getOnAction());
+                recentLibrariesBox.getChildren().add(recentLibraryLink);
+            }
+        }
```
@oops-shlok Is this fixed? I did not see "Outdated" markings, but a thumbs-up from calixtus.

koppor replied:
It is OK if this cannot be solved easily, then just comment accordingly.

koppor replied:
Ah, the fix is easy: add a "return" in the first branch - and remove `} else {`

oops-shlok replied:
yes! Actually not sure why the changes aren't there. I fixed it the other night when this review comment came up. Maybe some issue happened while pushing the code. I will push it again now. Thanks

koppor commented on src/main/java/org/jabref/gui/exporter/ExportCommand.java:
```diff
@@ -133,19 +133,23 @@ private void export(Path file, FileChooser.ExtensionFilter selectedExtensionFilt
-                                notificationPane.hide();
-                            })),
-                            Duration.seconds(5));
+                    boolean currentTabIsLibraryTab = tabSupplier.get() != null;
+                    // libraryTab is null in case of Welcome Tab
```
If possible, swap these. Typically, java comments come before the statement

koppor commented on src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -175,6 +177,9 @@ public JabRefFrame(Stage mainStage,
                 dialogService,
                 getOpenDatabaseAction());
+
+        fileHistory.disableProperty().bind(Bindings.isEmpty(fileHistory.getItems()));
```
Not sure if the `isEmpty()` could be propagated more down.

I feel that this is another issue which shouljd be tackled. -- Creating a follow-up issue is OK for me.

oops-shlok replied:
sure! Meanwhile I will try to propagate it.

koppor commented on src/main/java/org/jabref/gui/exporter/ExportCommand.java:
```diff
@@ -133,19 +133,23 @@ private void export(Path file, FileChooser.ExtensionFilter selectedExtensionFilt
+                    boolean currentTabIsLibraryTab = tabSupplier.get() != null;
+                    // libraryTab is null in case of Welcome Tab
+                    if (currentTabIsLibraryTab) {
+                        LibraryTab.DatabaseNotification notificationPane = tabSupplier.get().getNotificationPane();
```
Is this code ever called?

I mean - this is called when exporting a library - how can it hapen that the welcome tab is exported (!) successfully (!!)

oops-shlok replied:
Reverted this change

koppor commented on src/main/java/org/jabref/gui/WelcomeTab.java:
```diff
@@ -0,0 +1,281 @@
+
+        String url = switch (action) {
+            case HELP ->
+                    "https://help.jabref.org/";
```
Where does link come from? -- It should be `https://docs.jabref.org`.

koppor commented on src/main/java/org/jabref/gui/WelcomeTab.java:
```diff
@@ -0,0 +1,281 @@
+        String url = switch (action) {
+            case HELP ->
+                    "https://help.jabref.org/";
+            case OPEN_FORUM ->
+                    "https://discourse.jabref.org/";
```
Damn, this is constant duplication.

## First (crap) idea

Please add `getUrl` to `org.jabref.gui.edit.OpenBrowserAction` and then use this here.

(you can just check for `instanceof OpenBrowserAction openBrowserAction` and then use the URL

(does not work because you get standard action only)

## Second idea: introduce `org.jabref.gui.util.Urls.java`

Move URLs as `public static final` constants there.

Also move the URLs from org.jabref.gui.AbstractViewModel to there.

Then, there should be no dpulciate "discourse.jabref.org" string in JAbRef code base

oops-shlok replied:
Sure! will do it right away.

oops-shlok replied:
@koppor , shall I change for org/jabref/gui/frame/MainMenu.java as well?

oops-shlok replied:
Also, there isn't anything in org.jabref.gui.AbstractViewModel file.

package org.jabref.gui;

public class AbstractViewModel {
    // empty
}

koppor replied:
wrong copy paste

see screenshot 

![image](https://github.com/user-attachments/assets/3b5388c4-e313-4226-9b12-15da2e379c8d)

Note that you can search for an arbitrary string in the code using <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>F</kbd>. Maybe that helps to locate some code.

koppor replied:
> @koppor , shall I change for org/jabref/gui/frame/MainMenu.java as well?

Yes. There MUST NOT be two strings with the same URLs throughout GUI code.

oops-shlok commented:
@koppor Please take a look at the updated changes and lmk if its fine.

koppor commented on src/main/java/org/jabref/gui/WelcomeTab.java:
```diff
@@ -0,0 +1,274 @@
+    }
+
+    private Label createFooterLabel(String text) {
+        Label label = new Label(Localization.lang(text));
```
```suggestion
        Label label = new Label(text);
```

Localiziation was done at the caller

koppor reviewed (changes requested):
Localization test fails because of `Localization.lang(text)` was used. This call was not necessary.

koppor reviewed (changes requested):
localization missing

InAnYan commented on src/main/java/org/jabref/gui/WelcomeTab.java:
```diff
@@ -0,0 +1,274 @@
+        this.fileHistoryMenu = fileHistoryMenu;
+        this.buildInfo = buildInfo;
+
+        this.recentLibrariesBox = new VBox(5);
```
All over JabRef we often use 10 as spacing for `VBox`es and `HBox`es.

What if you could use it everywhere? How would it look?

I'm just feared of constant values

InAnYan replied:
Just like here: <https://github.com/JabRef/jabref/pull/12461/files#diff-83814543084e1de08c7978d3361eb0889b4c77aad0e1365a8a0d53b4484f2b75R87>

oops-shlok replied:
Sure! Will try it out and if it looks fine, will update it to 10.

InAnYan commented on src/main/java/org/jabref/gui/WelcomeTab.java:
```diff
@@ -0,0 +1,274 @@
+        container.setAlignment(Pos.CENTER);
+
+        Hyperlink onlineHelpLink = createFooterLink(Localization.lang("Online help"), StandardActions.HELP, IconTheme.JabRefIcons.HELP);
+        Hyperlink forumLink = createFooterLink(Localization.lang("Forum for support"), StandardActions.OPEN_FORUM, IconTheme.JabRefIcons.FORUM);
```
@koppor Forum is used not only for support.

Also for feature requests, introductions. 

What about changing to just "Forum"?

koppor replied:
Changed to "Community forum" by me

InAnYan commented on src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -350,8 +374,8 @@ private void initBindings() {
-            } else if (selectedTab == null) {
-                // All databases are closed
+            } else if (selectedTab == null || selectedTab instanceof WelcomeTab) {
+                // All databases are closed or Welcome Tab is open
```
Use `{@link ...}` here in the comment :)

koppor replied:
I implemented.

koppor commented on src/main/java/org/jabref/gui/WelcomeTab.java:
```diff
@@ -0,0 +1,274 @@
+        HBox container = new HBox(15);
+        container.setAlignment(Pos.CENTER);
+
+        Label versionLabel = new Label(Localization.lang("Current JabRef version") + ": " + buildInfo.version);
```
Sorry again

GOOD: Localization.lang("Current JabRef version: %0",  buildInfo.version);

Use parameters. Reason: even the `:` can be localized

InAnYan reviewed (changes requested):
Some small comments

calixtus commented:
I noticed, changelog entry is yet missing

oops-shlok commented:
> I noticed, changelog entry is yet missing

Added now.

koppor reviewed (dismissed):
No /// inline JavaComments.

The thing is that neither {@link} nor `[...]` is supported in Java comments.

I would propose to use `{@link ...}` (instead of Markdown links)

And please vote up https://youtrack.jetbrains.com/issue/IDEA-342361/Allow-JavaDocs-link-in-inline-comments - if possible

oops-shlok commented:
> No /// inline JavaComments.
> 
> The thing is that neither {@link} nor `[...]` is supported in Java comments.
> 
> I would propose to use `{@link ...}` (instead of Markdown links)
> 
> And please vote up https://youtrack.jetbrains.com/issue/IDEA-342361/Allow-JavaDocs-link-in-inline-comments - if possible

Sure!

koppor commented:
> > No /// inline JavaComments.
> > 
> > The thing is that neither {@link} nor `[...]` is supported in Java comments.
> > 
> > I would propose to use `{@link ...}` (instead of Markdown links)
> > 
> > And please vote up https://youtrack.jetbrains.com/issue/IDEA-342361/Allow-JavaDocs-link-in-inline-comments - if possible
> 
> Sure! Will upvote. 
> Also, @koppor on doing this, the checkstyle test fails. So, shall I proceed with that?

I don't get "that". Please copy and paste the checkstyle output here. I.e., try to make the question self-contained.

Background: I am trying to support mostly on my mobile phone for fast feedback. For a deep look I need my desktop machine and time for it. I cannot promise this for the next 10 days!

oops-shlok commented:
> > > No /// inline JavaComments.
> > > The thing is that neither {@link} nor `[...]` is supported in Java comments.
> > > I would propose to use `{@link ...}` (instead of Markdown links)
> > > And please vote up https://youtrack.jetbrains.com/issue/IDEA-342361/Allow-JavaDocs-link-in-inline-comments - if possible
> > 
> > 
> > Sure! Will upvote.
> > Also, @koppor on doing this, the checkstyle test fails. So, shall I proceed with that?
> 
> I don't get "that". Please copy and paste the checkstyle output here. I.e., try to make the question self-contained.
> 
> Background: I am trying to support mostly on my mobile phone for fast feedback. For a deep look I need my desktop machine and time for it. I cannot promise this for the next 10 days!

<img width="1680" alt="Screenshot 2025-03-19 at 12 27 16 AM" src="https://github.com/user-attachments/assets/4d844eed-e9b8-46b2-bac3-af44783a096f" />

koppor commented on src/main/java/org/jabref/gui/WelcomeTab.java:
```diff
@@ -0,0 +1,274 @@
+        container.setAlignment(Pos.CENTER);
+
+        Hyperlink onlineHelpLink = createFooterLink(Localization.lang("Online help"), StandardActions.HELP, IconTheme.JabRefIcons.HELP);
+        Hyperlink forumLink = createFooterLink(Localization.lang("Forum for support"), StandardActions.OPEN_FORUM, IconTheme.JabRefIcons.FORUM);
```
```suggestion
        Hyperlink forumLink = createFooterLink(Localization.lang("Forum"), StandardActions.OPEN_FORUM, IconTheme.JabRefIcons.FORUM);
```

koppor commented on src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -438,6 +463,34 @@ public void showLibraryTab(@NonNull LibraryTab libraryTab) {
     }
 
+    public void showWelcomeTab() {
+        // The loop iterates through all tabs in tabbedPane to check if a WelcomeTab already exists. If yes, it is selected
```
```suggestion
        // The loop iterates through all tabs in tabbedPane to check if a WelcomeTab already exists. If yes, it is selected.
```

oops-shlok commented:
@koppor Fixed the localisation issue. Anything else that needs to be changed?

koppor commented:
@oops-shlok I will finish the PR - I have time only today for a longer JabRef session - and I think, this back and forth is not worth our time. Time is better spend in new issues.

What I did

- Consistency in language: Help menu should have same strings than in the welcome tab. @InAnYan We changed both strings to "Community forum now"
- The welcome tab was shown if a file was opened. I fixed this by placing the call at a better place.

oops-shlok commented:
> @oops-shlok I will finish the PR - I have time only today for a longer JabRef session - and I think, this back and forth is not worth our time. Time is better spend in new issues.
> 
> What I did
> 
> * Consistency in language: Help menu should have same strings than in the welcome tab. @InAnYan We changed both strings to "Community forum now"
> * The welcome tab was shown if a file was opened. I fixed this by placing the call at a better place.

Sure @koppor! I am also available now. You can lmk in case anything needs to be done.

koppor commented:
Gui Tests are more fetcher tests - the remote seems to block - failing gui test is OK :)

calixtus commented:
Finally this pr is merged. @oops-shlok thanks for your addition. A great new feature for JabRef.

oops-shlok commented:
> Finally this pr is merged. @oops-shlok thanks for your addition. A great new feature for JabRef.

Thanks! It was great working on this feature. Will continue working on follow up PRs for this and also look into other issues. Thanks @calixtus and @koppor for the constant support.
